File: backend/parsers/instagram.py
from __future__ import annotations

from typing import Any
from urllib.parse import quote

from models import ProfileParseResult, ProfilePost
from parser_helpers import (
    extract_caption,
    extract_json_block,
    fetch_text,
    read_first_string,
    read_number,
    safe_json_parse,
    to_iso_date,
    truncate_posts,
)


def map_instagram_posts(user: dict[str, Any] | None) -> list[ProfilePost]:
    media = user.get("edge_owner_to_timeline_media") if user else None
    if not media:
        media = user.get("timeline_media") if user else None

    edges = media.get("edges", []) if isinstance(media, dict) else []
    posts: list[ProfilePost] = []

    for edge in edges:
        node = edge.get("node", {}) if isinstance(edge, dict) else {}
        caption_edges = node.get("edge_media_to_caption", {})
        caption_items = caption_edges.get("edges", []) if isinstance(caption_edges, dict) else []
        first_caption = None
        if caption_items and isinstance(caption_items[0], dict):
            first_caption = (caption_items[0].get("node") or {}).get("text")

        shortcode = read_first_string(node.get("shortcode"))
        post_id = read_first_string(node.get("id"), shortcode)
        if not post_id:
            continue

        posts.append(
            ProfilePost(
                id=post_id,
                url=f"https://www.instagram.com/p/{shortcode}/" if shortcode else None,
                caption=extract_caption(first_caption),
                publishedAt=to_iso_date(read_number(node.get("taken_at_timestamp"))),
            )
        )

    return truncate_posts(posts)


def map_instagram_user(user: dict[str, Any], normalized_url: str, username: str) -> ProfileParseResult:
    edge_followed_by = user.get("edge_followed_by") if isinstance(user.get("edge_followed_by"), dict) else {}
    edge_follow = user.get("edge_follow") if isinstance(user.get("edge_follow"), dict) else {}
    timeline = user.get("edge_owner_to_timeline_media") if isinstance(user.get("edge_owner_to_timeline_media"), dict) else {}

    return ProfileParseResult(
        platform="instagram",
        profileUrl=normalized_url,
        username=read_first_string(user.get("username"), username) or username,
        fullName=read_first_string(user.get("full_name")),
        bio=read_first_string(user.get("biography"), user.get("bio")),
        avatarUrl=read_first_string(user.get("profile_pic_url_hd"), user.get("profile_pic_url")),
        isPrivate=bool(user.get("is_private")),
        isVerified=bool(user.get("is_verified")),
        followersCount=read_number(edge_followed_by.get("count")),
        followingCount=read_number(edge_follow.get("count")),
        postsCount=read_number(timeline.get("count")),
        recentPosts=map_instagram_posts(user),
    )


def parse_via_api(normalized_url: str, username: str) -> ProfileParseResult | None:
    api_url = f"https://www.instagram.com/api/v1/users/web_profile_info/?username={quote(username)}"
    response_text = fetch_text(
        api_url,
        headers={
            "X-IG-App-ID": "936619743392459",
            "Referer": normalized_url,
        },
    )
    parsed = safe_json_parse(response_text)
    user = ((parsed or {}).get("data") or {}).get("user") if isinstance(parsed, dict) else None

    if not isinstance(user, dict):
        return None

    return map_instagram_user(user, normalized_url, username)


def parse_via_html(normalized_url: str, username: str) -> ProfileParseResult | None:
    page = fetch_text(normalized_url)
    block = extract_json_block(
        page,
        [
            r'<script type="application/json" data-sjs[^>]*>([\s\S]*?)</script>',
            r'"userInfo":(\{[\s\S]*?\}),"logging_page_id"',
        ],
    )

    if not block:
        return None

    embedded = safe_json_parse(block)
    user = None
    if isinstance(embedded, dict):
        user = (
            ((embedded.get("props") or {}).get("pageProps") or {}).get("userInfo") or {}
        ).get("user")

    if not isinstance(user, dict):
        wrapped = safe_json_parse(f'{{"userInfo":{block}}}')
        if isinstance(wrapped, dict):
            user = ((wrapped.get("userInfo") or {}).get("user"))

    if not isinstance(user, dict):
        return None

    return map_instagram_user(user, normalized_url, username)


def parse_instagram_profile(normalized_url: str, username: str) -> ProfileParseResult:
    attempts = [parse_via_api, parse_via_html]

    for attempt in attempts:
        try:
            parsed = attempt(normalized_url, username)
            if parsed:
                return parsed
        except Exception:
            continue

    raise RuntimeError("Instagram did not return public profile data. Public scraping can still be unstable.")

File: src/services/parser/instagram.ts
import { PlatformAdapter, ProfileParseResult, ProfilePost } from "../../types/parser";
import { normalizeProfileUrl } from "../../utils/url";
import {
  extractCaption,
  extractJsonBlock,
  fetchText,
  readFirstString,
  readNumber,
  safeJsonParse,
  toIsoDate,
  truncatePosts,
} from "./helpers";

type InstagramApiResponse = {
  data?: {
    user?: {
      biography?: string;
      edge_followed_by?: { count?: number };
      edge_follow?: { count?: number };
      edge_owner_to_timeline_media?: {
        count?: number;
        edges?: Array<{
          node?: {
            id?: string;
            shortcode?: string;
            taken_at_timestamp?: number;
            edge_media_to_caption?: {
              edges?: Array<{
                node?: { text?: string };
              }>;
            };
          };
        }>;
      };
      full_name?: string;
      hd_profile_pic_url_info?: { url?: string };
      is_private?: boolean;
      is_verified?: boolean;
      profile_pic_url_hd?: string;
      username?: string;
    };
  };
};

type InstagramEmbeddedData = {
  props?: {
    pageProps?: {
      userInfo?: {
        user?: Record<string, unknown>;
      };
    };
  };
};

function mapInstagramPosts(user: Record<string, unknown> | undefined): ProfilePost[] {
  const media = (user?.edge_owner_to_timeline_media ?? user?.timeline_media) as
    | {
        edges?: Array<{
          node?: Record<string, unknown>;
        }>;
      }
    | undefined;

  const edges = Array.isArray(media?.edges) ? media.edges : [];

  const posts: ProfilePost[] = [];

  for (const edge of edges) {
    const node = edge?.node;
    const captionEdges = node?.edge_media_to_caption as
      | { edges?: Array<{ node?: { text?: string } }> }
      | undefined;
    const firstCaption = captionEdges?.edges?.[0]?.node?.text;
    const shortcode = readFirstString(node?.shortcode);
    const id = readFirstString(node?.id, shortcode);

    if (!id) {
      continue;
    }

    posts.push({
      id,
      url: shortcode ? `https://www.instagram.com/p/${shortcode}/` : undefined,
      caption: extractCaption(firstCaption),
      publishedAt: toIsoDate(readNumber(node?.taken_at_timestamp)),
    });
  }

  return truncatePosts(posts);
}

function mapInstagramUser(user: Record<string, unknown>, normalizedUrl: string, username: string): ProfileParseResult {
  return {
    platform: "instagram",
    profileUrl: normalizedUrl,
    username: readFirstString(user.username) ?? username,
    fullName: readFirstString(user.full_name),
    bio: readFirstString(user.biography, user.bio),
    avatarUrl: readFirstString(user.profile_pic_url_hd, user.profile_pic_url),
    isPrivate: Boolean(user.is_private),
    isVerified: Boolean(user.is_verified),
    followersCount: readNumber((user.edge_followed_by as { count?: number } | undefined)?.count),
    followingCount: readNumber((user.edge_follow as { count?: number } | undefined)?.count),
    postsCount: readNumber((user.edge_owner_to_timeline_media as { count?: number } | undefined)?.count),
    recentPosts: mapInstagramPosts(user),
  };
}

async function parseViaApi(normalizedUrl: string, username: string) {
  const apiUrl = `https://www.instagram.com/api/v1/users/web_profile_info/?username=${encodeURIComponent(username)}`;
  const responseText = await fetchText(apiUrl, {
    headers: {
      "X-IG-App-ID": "936619743392459",
      Referer: normalizedUrl,
    },
  });
  const parsed = safeJsonParse<InstagramApiResponse>(responseText);
  const user = parsed?.data?.user;

  if (!user) {
    return null;
  }

  return mapInstagramUser(user as unknown as Record<string, unknown>, normalizedUrl, username);
}

async function parseViaHtml(normalizedUrl: string, username: string) {
  const page = await fetchText(normalizedUrl);
  const block = extractJsonBlock(page, [
    /<script type="application\/json" data-sjs[^>]*>([\s\S]*?)<\/script>/i,
    /"userInfo":(\{[\s\S]*?\}),"logging_page_id"/i,
  ]);

  if (!block) {
    return null;
  }

  const embedded = safeJsonParse<InstagramEmbeddedData>(block);
  const user =
    embedded?.props?.pageProps?.userInfo?.user ??
    (safeJsonParse<{ userInfo?: { user?: Record<string, unknown> } }>(`{"userInfo":${block}}`)?.userInfo?.user ??
      null);

  if (!user) {
    return null;
  }

  return mapInstagramUser(user, normalizedUrl, username);
}

export const instagramAdapter: PlatformAdapter = {
  canHandle(url) {
    return /instagram\.com/i.test(url);
  },
  async parseProfile(url) {
    const { normalizedUrl, username } = normalizeProfileUrl(url);

    const attempts = [() => parseViaApi(normalizedUrl, username), () => parseViaHtml(normalizedUrl, username)];

    for (const attempt of attempts) {
      try {
        const parsed = await attempt();

        if (parsed) {
          return parsed;
        }
      } catch {
        continue;
      }
    }

    throw new Error(
      "Instagram не отдал публичные данные. Для этой платформы без бэкенда и авторизованного API парсинг может быть нестабильным.",
    );
  },
};
